### ISSUE TITLE
What type to use for function that should work with data from different queries?

### ISSUE THREAD

**Issue by paoloposo** [comigor/artemis#271]
Hi, I'm just getting started with Artemis and I cannot wrap my head around one specific concept.

I'm using Artemis for generating Dart classes from the schema and then the [graphql](https://github.com/zino-app/graphql-flutter) client to make the request. That gives me a `Map<String, dynamic>` as a result which contains my data in denormalized form.

For example, when using the [Pokemon API](https://graphql-pokemon2.vercel.app) and the two sample queries (expand code below), I get two classes, `SinglePokemonQuery` and `MultiplePokemonQuery` which I can instantiate and then access the `document` property on. This is what I pass into the client and I get the denormalized data `Map` as a result.

<details>
<summary>Full code example</summary>
  
**Dart code**
```dart
// Initialize client
final link = HttpLink('https://graphql-pokemon2.vercel.app');
final client = GraphQLClient(link: link, cache: GraphQLCache());

// Make request for a single Pokemon
final singleResult = await client.query(QueryOptions(
  document: SinglePokemonQuery().document,
  variables: SinglePokemonArguments(name: 'Squirtle').toJson(),
));
print(SinglePokemon$Query.fromJson(singleResult.data).pokemon.name);

// Make request for multiple Pokemon
final multipleResult = await client.query(QueryOptions(
  document: MultiplePokemonQuery().document,
));
print(MultiplePokemon$Query.fromJson(multipleResult.data).pokemons.first.name);
```

**GraphQL Queries**
```graphql
query SinglePokemon($name: String) {
    pokemon(name: $name) {
        name
        number
    }
}

query MultiplePokemon {
    pokemons(first: 3) {
        name
        number
        evolutions {
            name
            number
        }
    }
}
```
GraphQL schema can be found at [https://graphql-pokemon2.vercel.app](https://graphql-pokemon2.vercel.app)
</details>

To access the data with type support, I can now use the `fromJson` methods of the respective generated response class (`SinglePokemon$Query` or `MultiplePokemon$Query`). I can then traverse the structure and access the Pokemon data.

```dart
SinglePokemon$Query.fromJson(singleResult.data).pokemon.runtimeType // SinglePokemon$Query$Pokemon
MultiplePokemon$Query.fromJson(multipleResult.data).pokemons.first.runtimeType // MultiplePokemon$Query$Pokemon
```
However I now have two Dart types (`SinglePokemon$Query$Pokemon` and `MultiplePokemon$Query$Pokemon`) which both represent the same GraphQL type (`Pokemon`). I understand why this is necessary and makes sense. Although returning the same GraphQL types, the queries might contain a different set of fields and thus they need to have their own separate response classes.

This brings me to my question:
**How would I write a function that should work with any object of a specific GraphQL type, regardless of the query it came from?**

For example, if I want to have a function like so:

```dart
// what type can I use here instead of dynamic?
void doSomethingWithPokemon(dynamic pokemon) {
  // print(pokemon.name);
  // ...
}
```

I would have to use a dynamic type parameter, since theres is no `Pokemon` type generated, and so i cannot take advantage of Dart's type system.

Is there a solution for this that I am mising?

**Comment by comigor** [comment]
Hey @paoloposo!

First of all, thanks for drilling down at the core question and having those examples to back offit off, it really help to understand!

So yeah, currently each `SchemaMap` will generated its own set of classes to a specific (single file) output. If you're declaring two different `SchemaMap`, for the reasons you've stated above, you'll have two different classes (maybe with a different set of fields).

If you want to correlate them while still using this library, you'd have to "merge" them somehow, and currently this is only possible using fragments. Once you define a fragment, it can be used in more than one query (by globbing more than one query in the same `queries_glob` configuration). By doing that (eg. defining a fragment `PokemonFrag` with only the `name` field) both generated classes would extend it and could be used in your `doSomethingWithPokemon` function.

On the other hand, to be completely generic, you could use `JsonSerializable` class, as all generated classes extend from it but, as you said, it won't be typed anymore.

**Comment by vasilich6107** [comment]
Extracting common entities into fragments should result in `FragmentNameMixin` generations.
Then you'll be able to use `FragmentNameMixin` as type in your project.

**Comment by paoloposo** [comment]
Thanks very much guys, with your help I could solve my problem! Using GraphQL fragments indeed did the trick. I ran into the same problem as #52 first (getting the `Bad state: No element` error), but I think specifying the `fragments_glob` field in the `build.yaml` file fixed that. Also your own Pokemon example in the repo was really helpful (didn't even know you had this when I opened the issue). As @vasilich6107 stated, the mixin classes were generated and these can be used as parameter type.
